app/controllers/lesson_completions_controller.rb

```diff
@@ -1,51 +1,43 @@
 class LessonCompletionsController < ApplicationController
   
-  before_filter :authenticate_request
+  before_action :authenticate_request
+  before_action :lookup_lesson
   
   def create
-    # Validate that the id does in fact correspond to a real lesson
-    # Note that the authentication before_filter makes sure we've got a current_user
-    @lesson = Lesson.where(:id => params[:lesson_id]).first
-
-    if @lesson.nil?
-      render :nothing => true, :status => 400 # bad request
-    else   
-      @course = @lesson.course
-      @next_lesson = @lesson.next_lesson
-      lc = LessonCompletion.new(:student_id => current_user.id, :lesson_id => @lesson.id)
-      if lc.save
-        render "create", :formats => [:js]
-      else
-        render :nothing => true, :status => 400 # bad request
-      end
+    @course = @lesson.course
+    @next_lesson = @lesson.next_lesson
+    lc = LessonCompletion.new(:student_id => current_user.id, :lesson_id => @lesson.id)
+    if lc.save
+      render :create, formats: [:js]
+    else
+      head :bad_request
     end
   end
   
   def destroy
-    # Validate that the id does in fact correspond to a real lessons
-    # Note that the authentication before_filter makes sure we've got a current_user
-    @lesson = Lesson.where(:id => params[:lesson_id]).first
-
-    if @lesson.nil?
-      render :nothing => true, :status => 400 # bad request
-    else   
-      @course = @lesson.course
-      @next_lesson = @lesson.next_lesson
-      lc = LessonCompletion.where(:student_id => current_user.id, :lesson_id => @lesson.id).first
-      if lc.nil? 
-        render :nothing => true, :status => 400 # bad request
-      elsif lc.delete
-        render "create", :formats => [:js] 
-      else
-        render :nothing => true, :status => 400 # bad request
-      end
+    @course = @lesson.course
+    @next_lesson = @lesson.next_lesson
+    lc = LessonCompletion.where(:student_id => current_user.id, :lesson_id => @lesson.id).
+                          first
+    if lc.nil?
+      head :bad_request
+    else
+      lc.destroy
+      render :create, formats: [:js]
     end
   end
 
   private
-    def authenticate_request
-      unless user_signed_in?
-        render :nothing => true, :status => 401 # unauthorized
+
+    def lookup_lesson
```

Comment:
This seems like defensive programming to me. Because this controller will be called from a lesson we can safely assume the lesson will always exist.  We can simplify this method to:

```
def lookup_lesson
  @lesson = Lesson.find(params[:lesson_id])
end
```